src/cr/cube/slices.py

```diff
@@ -0,0 +1,1140 @@
+# encoding: utf-8
+
+from __future__ import division
+
+import numpy as np
+from scipy.stats.contingency import expected_freq
+
+from cr.cube.util import lazyproperty
+from cr.cube.enum import DIMENSION_TYPE as DT
+
+
+class _CatXCatSlice(object):
+    def __init__(self, dimensions, counts, base_counts):
+        self._dimensions = dimensions
+        self._counts = counts
+        self._base_counts = base_counts
+
+    @lazyproperty
+    def _row_dimension(self):
+        return self._dimensions[0]
+
+    @lazyproperty
+    def _column_dimension(self):
+        return self._dimensions[1]
+
+    @lazyproperty
+    def _row_elements(self):
+        return self._row_dimension.valid_elements
+
+    @lazyproperty
+    def _column_elements(self):
+        return self._column_dimension.valid_elements
+
+    @lazyproperty
+    def _row_generator(self):
+        return zip(self._counts, self._base_counts, self._row_elements)
+
+    @lazyproperty
+    def _column_generator(self):
+        return zip(self._counts.T, self._base_counts.T, self._column_elements)
+
+    @lazyproperty
+    def rows(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._row_generator
+        )
+
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._column_generator
+        )
+
+    @lazyproperty
+    def table_margin(self):
+        return np.sum(self._counts)
+
+
+class _1DCatSlice(_CatXCatSlice):
```

Comment:
I don't know about you guys, but I find this rather elegant. The need for it arose when I was dealing with exporter tests. I was able to adopt the vector-based logic of the 2D slice to a 1D slice in a very short amount of time.

Reply:
Yeah, I'm not sure, but I really like that it works so far and that totally makes it what we should do now in my mind. The first objective has to be to get to a working version that passes all the tests. This is _not_ so we can declare victory and move on, but because refactoring from there is such a more predictable and low-risk proposition. Any changes made with a full working set of integration tests is easy to see any breakage and it makes incremental refinement productive and fun. And, you can deliver at any moment if that turns out to be necessary, the critical thing being a stable external interface.

On the nature of a 1D cube result, I'm not convinced that thinking of it as a variant of a 2D matrix is the right long-term design. The thing is that a 1D item (we'll say "strand" for now) _has no "body" columns_. In particular, it doesn't "just have 1 column". It _also_ doesn't have any _row_values_ of the type that appear in a 2D slice. It does have a dimension, type, and counts, but I'm inclined to think the "column" of values it has is really a _margin_. So like a 2D slice _without the body_. It can still have _summary-columns_, but by nature cannot have any body columns (or body rows of course, since those are two sides of the same coin).

A confirming view is that a 2D/slice represents a _crosstab_ and a 1D/strand _does not_.

So I'm inclined to think this means a _strand_ should be a class hierarchy distinct from _slice_ (although perhaps some shared base class is possible).

All of this can wait until the refactoring step, but just some food for thought.